cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,409 @@ struct ReplaceWithMaskFunctor {
+                             ArrayData last_valid_value_chunk,
+                             int64_t* last_valid_value_offset) {
+  uint8_t* out_bitmap = output->buffers[0]->mutable_data();
+  uint8_t* out_values = output->buffers[1]->mutable_data();
+  arrow::internal::CopyBitmap(current_chunk.buffers[0]->data(), current_chunk.offset,
+                              current_chunk.length, out_bitmap, output->offset);
+  ReplaceWithMask<Type>::CopyData(*current_chunk.type, out_values,
+                                  /*out_offset=*/output->offset, current_chunk,
+                                  /*in_offset=*/0, current_chunk.length);
 
-const FunctionDoc replace_with_mask_doc(
-    "Replace items selected with a mask",
-    ("Given an array and a boolean mask (either scalar or of equal length),\n"
-     "along with replacement values (either scalar or array),\n"
-     "each element of the array for which the corresponding mask element is\n"
-     "true will be replaced by the next value from the replacements,\n"
-     "or with null if the mask is null.\n"
-     "Hence, for replacement arrays, len(replacements) == sum(mask == true)."),
-    {"values", "mask", "replacements"});
+  bool has_fill_value = *last_valid_value_offset != -1;
+  int64_t write_offset = direction == 1 ? 0 : current_chunk.length - 1;
+  int64_t bitmap_offset = 0;
 
-void RegisterVectorReplace(FunctionRegistry* registry) {
-  auto func = std::make_shared<VectorFunction>("replace_with_mask", Arity::Ternary(),
-                                               &replace_with_mask_doc);
+  arrow::internal::OptionalBitBlockCounter counter(null_bitmap, output->offset,
+                                                   current_chunk.length);
+
+  while (bitmap_offset < current_chunk.length) {
+    BitBlockCount block = counter.NextBlock();
+    if (block.AllSet()) {
+      *last_valid_value_offset =
+          write_offset + direction * (block.length - 1 + bitmap_offset);
+      has_fill_value = true;
+      last_valid_value_chunk = current_chunk;
+    } else {
+      uint64_t block_start_offset = write_offset + direction * bitmap_offset;
+      uint64_t write_value_offset = block_start_offset;
+      for (int64_t i = 0; i < block.length; i++, write_value_offset += direction) {
+        auto current_bit = bit_util::GetBit(null_bitmap, bitmap_offset + i);
+        if (!current_bit) {
+          if (has_fill_value) {
+            ReplaceWithMask<Type>::CopyData(*current_chunk.type, out_values,
+                                            write_value_offset, last_valid_value_chunk,
+                                            *last_valid_value_offset,
+                                            /*length=*/1);
+            bit_util::SetBitTo(out_bitmap, write_value_offset, true);
+          }
+        } else {
+          has_fill_value = true;
+          *last_valid_value_offset = write_value_offset;
+          last_valid_value_chunk = current_chunk;
+        }
+      }
+    }
```

Comment:
Note that if block.popcount is 0, it may still be worth duplicating this loop and removing the bitmap check and `has_fill_value` check. I was just saying that it is not worth allocating a Scalar for that case.